glib/src/subclass/signal.rs

```diff
@@ -526,10 +538,29 @@ impl Signal {
             SignalRegistration::Registered { .. } => unreachable!(),
         };
 
+        let return_type = self.return_type;
+
         let class_handler = class_handler.map(|class_handler| {
             Closure::new(move |values| unsafe {
                 let instance = gobject_ffi::g_value_get_object(values[0].to_glib_none().0);
-                class_handler(&SignalClassHandlerToken(instance as *mut _), values)
+                let res = class_handler(&SignalClassHandlerToken(instance as *mut _, return_type.into(), values.as_ptr()), values);
+
+                if return_type == Type::UNIT {
+                    if let Some(ref v) = res {
+                        assert!(res.is_none(), "Signal has no return value but class handler returned a value of type {}", v.type_());
```

Comment:
How can this `assert` be triggered since we checked just before `res` was `Some`?

Reply:
It will always hit, that's how :)

I went with an assert here so it's a bit clearer from the error what went wrong. It could also just be a plain `panic!`.

Reply:
Well, my comment makes it very obvious that the code isn't clear. Also, just realized that you match on `res` just after. Might be better to match directly on it and `panic` on the `Some` branch. That'll remove one extra check.

Reply:
The check is optimized away. The difference is in the output: just a plain panic with that message, vs. a `assertion res.is_none() failed: that message`. It seems more helpful to the user, but if you think the code is too hard to read otherwise I can change that.

Reply:
The code is very convoluted here. Please change it. :)